scenario_player/main.py

```
@@ -220,94 +224,158 @@ def run(
+                ui = ScenarioUIManager(runner, log_buffer, log_file_name, success)
+            else:
+                ui = nullcontext()
+            with ui:
+                runner.run_scenario()
+    except Exception:
+        raise
+
+
+class ScenarioUIManager:
+    def __init__(self, runner, log_buffer, log_file_name, success):
+        self.ui = ScenarioUI(runner, log_buffer, log_file_name)
+        self.success = success
+        self.ui_greenlet = self.ui.run()
+
+    def __enter__(self):
+        return self.success
+
+    def __exit__(self, type, value, traceback):
         try:
-            if enable_ui and ui:
-                ui.set_success(success)
-                log.warning("Press q to exit")
-                while not ui_greenlet.dead:
-                    gevent.sleep(1)
-            service_process.stop()
-        except ServiceProcessException:
-            service_process.kill()
+            self.ui.set_success(self.success.get())
+            log.warning("Press q to exit")
+            while not self.ui_greenlet.dead:
+                gevent.sleep(1)
         finally:
-            runner.node_controller.stop()
-            if ui_greenlet is not None and not ui_greenlet.dead:
-                ui_greenlet.kill(ExitMainLoop)
-                ui_greenlet.join()
-            exit(exit_code)
+            if self.ui_greenlet is not None and not self.ui_greenlet.dead:
+                self.ui_greenlet.kill(ExitMainLoop)
+                self.ui_greenlet.join()
+
+
+class ServiceProcessManager:
+    def __init__(self):
+        self.service_process = ServiceProcess()
+        self.service_process.start()
+
+    def __enter__(self):
+        return self.service_process
+
+    def __exit__(self, type, value, traceback):
+        try:
+            self.service_process.stop()
+        except ServiceProcessException:
+            log.exception("ServiceProcessManager died")
+        except Exception:
+            self.service_process.kill()
```

Is it right to swallow the exception here?

---

I'll log more verbose, just to be sure.